src/publish-build/index.ts

```diff
@@ -0,0 +1,97 @@
+import * as core from '@actions/core';
+import {
+  S3Client,
+  PutObjectCommand,
+  PutObjectCommandOutput,
+  PutObjectCommandInput,
+} from '@aws-sdk/client-s3';
+import { promises as fs, createReadStream } from 'fs';
+import mime from 'mime';
+import * as path from 'path';
+
+const REQUIRED_ENV_VARS = [
+  'AWS_BUCKET',
+  'PROJECT',
+  'SOURCE_DIR',
+  'DESTINATION_DIR',
+  'COMMIT_HASH',
+  'GITHUB_WORKSPACE',
+];
+
+interface envParams {
+  AWS_BUCKET: string;
+  PROJECT: string;
+  SOURCE_DIR: string;
+  DESTINATION_DIR: string;
+  COMMIT_HASH: string;
+  GITHUB_WORKSPACE: string;
+}
+
+function getEnvVars(): envParams {
+  const res: { [key: string]: string } = {};
+  for (const requiredVar of REQUIRED_ENV_VARS) {
+    if (!process.env[requiredVar]) {
+      const errMsg = `Required env variable missing: ${requiredVar}`;
+      core.error(errMsg);
+      throw new Error(errMsg);
+    }
+    res[requiredVar] = process.env[requiredVar] ?? '';
```

Comment:
nit: I wonder if this nullish coalescer is necessary because it would have thrown an error?

Reply:
Here's a way that this can be done without running into TS errors. There are a few more moving parts, but this should make TypeScript happy. I added some explanations, but if I can clarify something more let me know!!

```ts
const REQUIRED_ENV_VARS = [
  'AWS_BUCKET',
  'PROJECT',
  'SOURCE_DIR',
  'DESTINATION_DIR',
  'COMMIT_HASH',
  'GITHUB_WORKSPACE',
] as const;

// translates the array of values in REQUIRED_ENV_VARS into a literal type i.e. RequiredVar = 'AWS_BUCKET' | 'PROJECT' ....
type RequiredVar = (typeof REQUIRED_ENV_VARS)[number];

// We can then use this literal type to create an object definition. Have to use a type here instead of an interface
// as interfaces don't support mapped types very well :/
// This boils down to the same intreface, but with the added benefit of being able to assign these string values to the object in
// the getEnvVars function. If this is not done, TypeScript sees a string versus a key within the `envParams` type. Because of this,
// an error will occur if you try assigning a value to the envParams object.
type envParams = {
  [v in RequiredVar]: string;
};

function getEnvVars(): envParams {
  const res: Partial<envParams> = {};
  for (const requiredVar of REQUIRED_ENV_VARS) {
    const envVar = process.env[requiredVar];
    if (!envVar) {
      const errMsg = `Required env variable missing: ${requiredVar}`;
      core.error(errMsg);
      throw new Error(errMsg);
    }
    res[requiredVar] = envVar;
  }
  return res as envParams;
}
```

Reply:
omg grosssss typescript!! 😭 @branberry you right, you always right!

Reply:
Lol, I def ain't always right... I've just had to do this so many times already hahaha. I've been wallowing in the murky TypeScript underbelly for far too long...

Reply:
this part was throwing an error so i updated it to be a `type Record<RequiredVar, string>;`

```
type envParams = {
  [var in RequiredVar]: string;
};
```

Reply:
Ah, the issue is that `var` is a JavaScript keyword.. If you use something like `V` instead of `var` the error goes away.

But, I prefer the `Record` as that is more readable